### ISSUE TITLE
MD063 Title-Case Corrupts Ordinals Preceded by Punctuation (`(2nd` → `(2Nd`)

### ISSUE THREAD

**Issue by StevenACoffman** [rvben/rumdl#848]
### rumdl version

0.2.64

### Installation method

Homebrew

### Bug description

# MD063 Title-Case Corrupts Ordinals Preceded by Punctuation (`(2nd` → `(2Nd`)

## Environment

- rumdl **0.2.64** (also reproduced on 0.2.49 — unchanged across those versions)
- macOS 26.5.1, arm64 (Homebrew)

## Summary

MD063 (`heading-capitalization`, `style = "title-case"`) correctly leaves ordinals
(`1st`, `2nd`, `3rd`, `21st`, `103rd`) alone when they follow whitespace, and even
*repairs* a mis-cased one back to lowercase. But when an ordinal is immediately preceded
by punctuation — `(`, `[`, `{`, `<`, `"`, `'`, `,`, `/`, `—` — the ordinal is not
recognised and its letters are title-cased, producing `2Nd`, `3Rd`, `1St`.

`rumdl fmt` writes the corruption, and because the corrupted form is itself stable in
that position, **it never self-heals**. `rumdl check` reports the same thing as a
suggested fix:

```text
c.md:3:4: [MD063] Heading should use title case: 'paren (2nd item)' -> 'Paren (2Nd Item)' [*]
```

## Minimal Reproduction

`.rumdl.toml`:

```toml
[MD063]
enabled = true
style = "title-case"
```

`t.md`:

```markdown
# Heading

## plain 2nd item

## paren (2nd item)
```

Run `rumdl fmt t.md`. Result:

```markdown
## Plain 2nd Item          <- correct
## Paren (2Nd Item)        <- expected "(2nd Item)"
```

### Expected behavior

Ordinals should be preserved (and repaired) when surrounded by any non-alphanumeric (space and/or punctuation)
```markdown
## Plain 2nd Item 
## Paren (2nd Item)
```

### Actual behavior

The second ordinal example is mangled below:
```markdown
## Plain 2nd Item
## Paren (2Nd Item) 
```

### Configuration

```toml
[MD063]
enabled = true
style = "title-case"
```

### Operating system

macOS

## Which Preceding Characters Trigger It

Same ordinal, same heading, only the preceding character differs:

| Preceding character | Input                   | Output                  |         |
| ------------------- | ----------------------- | ----------------------- | ------- |
| (line start)        | `2nd at line start`     | `2nd at Line Start`     | ok      |
| space               | `plain 2nd item`        | `Plain 2nd Item`        | ok      |
| `-`                 | `hyphen -2nd item`      | `Hyphen -2nd Item`      | ok      |
| `(`                 | `paren (2nd item)`      | `Paren (2Nd Item)`      | **bug** |
| `[`                 | `bracket [2nd item]`    | `Bracket [2Nd Item]`    | **bug** |
| `{`                 | `brace {2nd item}`      | `Brace {2Nd Item}`      | **bug** |
| `<`                 | `angle <2nd item>`      | `Angle <2Nd Item>`      | **bug** |
| `"`                 | `quote "2nd item"`      | `Quote "2Nd Item"`      | **bug** |
| `'`                 | `apostrophe '2nd item'` | `Apostrophe '2Nd Item'` | **bug** |
| `,`                 | `comma ,2nd item`       | `Comma ,2Nd Item`       | **bug** |
| `/`                 | `slash /2nd item`       | `Slash /2Nd Item`       | **bug** |
| `—`                 | `emdash —2nd item`      | `Emdash —2Nd Item`      | **bug** |

## The Tool Both Produces and Repairs the Same Token

The two lines differ only in the character before the ordinal:

```markdown
## already mangled 2Nd item          ->  ## Already Mangled 2nd Item   (repaired)
## paren already mangled (2Nd item)  ->  ## Paren Already Mangled (2Nd Item)   (left corrupt)
```

MD063 evidently knows `2Nd` is wrong, since it fixes it in the first case. It emits that
exact form in the second.

## Scope: Ordinals Only

Other token classes are *not* affected by a leading parenthesis, so this is specific to
ordinal handling rather than a general tokenisation problem:

```markdown
## cased word parened (SkillLens here)  ->  ## Cased Word Parened (SkillLens Here)   ok
## acronym parened (NASA rules)         ->  ## Acronym Parened (NASA Rules)          ok
## minor word parened (a tale of two)   ->  ## Minor Word Parened (A Tale of Two)    ok (phrase start)
```

## Likely Cause and Suggested Fix

Consistent with the observations, though I have not read the source: the ordinal test
looks to be applied to the whole whitespace-delimited token *including* attached
punctuation, so an anchored match such as `^\d+(st|nd|rd|th)$` fails on `(2nd` and the
token falls through to "capitalise the first letter". Hyphen-preceded ordinals survive,
which suggests hyphenated tokens are already split before the test runs.

Stripping leading and trailing punctuation before the ordinal test — matching whatever
already handles the hyphen case — would fix all of the rows above.

## Real-World Impact

Found in a repository where `rumdl fmt` silently corrupted a committed heading:

```diff
-### Experiment Adjudication (2nd Consumer: adh `verdict`, 2026-08-04)
+### Experiment Adjudication (2Nd Consumer: Adh `verdict`, 2026-08-04)
```

The corrupted line has now survived several formatting passes, because `(2Nd` is stable
in that position. A wrapping code span is the only spelling that survives — `` (`2nd` ``
is left alone — which is a workaround rather than a fix.

**Comment by rvben** [comment]
Thanks @StevenACoffman. `is_numeric_ordinal` scanned the token before trimming its wrapping punctuation, so `(2nd` never matched and got recased to `(2Nd`.

One correction to the report while measuring it: this is title case only. Sentence case lowercases the words it recases, so `(2nd)` came through it unharmed. Also worth knowing if you have already run the old version over a repo: `(2Nd)` is now reported and restored to `(2nd)`, so documents it already corrupted get repaired rather than staying silently wrong. Fixed in v0.2.66.
